---
title: "Introduction"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This package is crafted to provide a collection of helper functions that aid in bioinformatics analysis.It does not perform independent analyses. Feel free to contact me if you have any requests for new functions or suggestions for improvement.

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(easybio)
```

## Annotate Single Cell Clusters Using the CellMarker2.0 Database

```{r}
head(pbmc.markers)
```

```{r}
(marker <- matchCellMarker2(marker = pbmc.markers, n = 50, spc = 'Human')[, head(.SD, 2), by=cluster])
```

Behind this function:

- First, it matches the top 50 genes in each cluster with the CellMarker2 database.
- Next, it calculates the cell count for each matched cell type and prints the result to help you identify the possible cell type for the cluster.

Visualize marker dot plots for similar clusters:

```{r,eval=FALSE}
cls <- list(
  c(1, 5, 7), 
  c(8),
  c(3),
  c(0,2, 4, 6)
)
plotSeuratDot(seuratObject, cls, marker = marker, n = 50, spc = 'Human', topcellN = 2)
```

Behind this function:

- It searches for potential cell markers from the result of matchCellMarker2.
- Next, it invokes `Seurat::DotPlot` to generate the corresponding dot plot. The resulting plot will be saved in the directory ‘tmp_check_marker’.

Construct the named vector for your annotation:

```{r}
cl2cell <- finsert(
  expression(
  c(1, 5) == "Monocyte",
  c(7) == "DC",
  c(8) == "megakaryocyte",
  c(3) == "B.cell",
  c(0, 2) == "Naive.CD8.T.cell",
  c(4) == "Cytotoxic.T.Cell",
  c(6) == "Natural.killer.cell",
), len = 9)
cl2cell
```

You can also directly obtain markers:

```{r}
get_marker(spc = 'Human', cell = c('Monocyte', 'Neutrophil'), number = 5, min.count = 1)
```

#### Downloading GEO Data

To download the GEO dataset, please utilize the `download_geo` function.

```r
x <- download_geo('gseid')
```

This function will yield a list containing count data, sample information, and gene data.

#### limma-voom Workflow for RNA-Seq Analysis

Three functions have been crafted for this workflow.

```r
x <- dgeList(count, sample.info, feature.info)
x <- dprocess.DGEList(x, group.column, min.count)
x <- limmaFit(x, group.column)
```

For a comprehensive understanding of this process, refer to the article [RNA-seq analysis is as easy as 1-2-3 with limma, Glimma and edgeR](https://doi.org/10.12688/f1000research.9005.3).

Subsequently, visualize the differentially expressed genes using the `plotVolcano` function.

```r
# Consult the help page to familiarize yourself with the arguments.
?plotVolcano

```

## Gene Set Enrichment Analysis (GSEA)

The core function is derived from the fgsea package, with slight modifications applied to enhance its visual appeal.

```{r}
library(fgsea)
library(easybio)

data(examplePathways)
data(exampleRanks)
```

Execute the fgsea analysis.

```{r,fig.width=4, fig.height=2.7}
fgseaRes <- fgsea(pathways = examplePathways, 
                  stats    = exampleRanks,
                  minSize  = 15,
                  maxSize  = 500)
plotGSEA(
  fgseaRes, 
  pathways = examplePathways, 
  pwayname = "5991130_Programmed_Cell_Death", 
  stats = exampleRanks, 
  save = TRUE
)
```

To achieve optimal visualization, the plot is saved for review.

## Over-Representation Analysis

Perform the Over-Representation Analysis (ORA).

```{r}
foraRes <- fora(examplePathways, genes=tail(names(exampleRanks), 200), universe=names(exampleRanks))
```

Examine the results.

```{r, fig.width=8, fig.height=2}
# Adjust the pathway position on the y-axis based on the adjusted p-value (padj)
foraRes[, pathway := factor(pathway, levels = rev(pathway))]

plotORA(data = foraRes[1:8], x = -log10(padj), y = pathway, size = overlap, fill = 'constant')
```
